pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -215,8 +218,8 @@ func (rc *reconciler) reconcile(ctx context.Context) {
 				logger.Error(err, "Failed to get health of node", "node", klog.KRef("", string(attachedVolume.NodeName)))
 			}
 
-			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration.
-			forceDetach := !isHealthy && timeout
+			// Force detach volumes from unhealthy nodes after maxWaitForUnmountDuration if force detach is enabled
+			forceDetach := !isHealthy && timeout && !rc.disableForceDetachOnUnmount
```

Comment:
does the disableForceDetach bit need to inform `verifySafeToDetach := !(timeout || hasOutOfServiceTaint)` below as well?

Reply:
It wouldn't make it past L231 but it won't hurt to add in case this function ever gets refactored in the future

Reply:
Whether to trigger detach and whether to verifySafeToDetach can be affected by 5 variables:
isHealthy | timeout | disableForceDetachOnUnmount | MountedByNode | hasOutOfServiceTaint

I feel it brings some trouble when reasoning about all the conditions. Trying to think of a way to better express this.

```
verifySafeToDetach = true
if hasOutOfServiceTaint {
       // this is a sufficient condition to continue with detaching without verification no matter any other conditions
       verifySafeToDetach = false
} else if  timeout & !isHealthy & !disableForceDetachOnTimeOut {
       // it is already timeout and node is NOT healthy, and forceDetach is NOT disabled, force detach can happen
      verifySafeToDetach = false
} else if !MountedByNode {
       // it is no longer mounted, detach can be triggered.
} else {
     // all other condition, we should skip detach
     continue
}

... detach operation begins ...

```

Reply:
I've made the following changes based on the feedback here:
* Added a logging message for when the timeout expires but is suppressed by the flag.
* Updated the semantics of how the `timeout` variable is calculated; it now includes `rc. disableForceDetachOnTimeOut`. This is because when I added the unit test to bypass this flag with the node taint, we were updating the wrong counter in the code below (the detach was counting towards `metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonTimeout)` and not `metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonOutOfService)`). This means that `rc. disableForceDetachOnTimeOut` is also included in `verifySafeToDetach`.
* @jingxu97: The more that I stare at this code, the more I think your refactor makes sense. Is this something I can pursue as a follow-up change?

Reply:
lgtm, thank you!

Reply:
Marking this as resolved; will follow up later.